File: python/leetcode/search-in-a-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def searchBST(self, root, val):
        """
        :type root: TreeNode
        :type val: int
        :rtype: TreeNode
        """
        return self.search(root, val)
    
    def search(self, root, val):
        if root.val == val:
            return root
        if val < root.val and root.left is not None:
            return self.search(root.left, val)
        if val > root.val and root.right is not None:
            return self.search(root.right, val)

File: java/leetcode/search-in-a-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode searchBST(TreeNode root, int val) {
        return search(root, val);
    }
    
    private TreeNode search(TreeNode root, int val) {
        if (root.val == val) {
            return root;
        } else if (val < root.val && root.left != null) {
            return search(root.left, val);
        } else if (val > root.val && root.right != null) {
            return search(root.right, val);
        }
        return null;
    }
}
